packages/core/src/CreateNodeAsTool.ts

```diff
@@ -0,0 +1,262 @@
+ * @param {string[]} matchingKeys - Array of matching keys.
+ * @param {Record<string, string>} values - Object containing values for the keys.
+ * @returns {Record<string, unknown>} The built nested object structure.
+ */
+function buildStructureFromMatches(
+	baseKey: string,
+	matchingKeys: string[],
+	values: Record<string, string>,
+): Record<string, unknown> {
+	const result = {};
+
+	for (const matchingKey of matchingKeys) {
+		const decodedKey = decodeDotNotation(matchingKey);
+		// Extract the part of the key after the base key
+		const remainingPath = decodedKey
+			.slice(baseKey.length)
+			.split('.')
+			.filter((k) => k !== '');
+		let current: Record<string, unknown> = result;
+
+		// Build the nested structure
+		for (let i = 0; i < remainingPath.length - 1; i++) {
+			if (!(remainingPath[i] in current)) {
+				current[remainingPath[i]] = {};
+			}
+			current = current[remainingPath[i]] as Record<string, unknown>;
+		}
+
+		// Set the value at the deepest level
+		const lastKey = remainingPath[remainingPath.length - 1];
+		current[lastKey ?? matchingKey] = values[matchingKey];
+	}
+
+	// If no nested structure was created, return the direct value
+	return Object.keys(result).length === 0 ? values[encodeDotNotation(baseKey)] : result;
+}
+
+/**
+ * Extracts the description from a placeholder string.
+ * @param {string} value - The placeholder string.
+ * @returns {string} The extracted description or a default message.
+ */
+function extractPlaceholderDescription(value: string): string {
+	const match = value.match(/{{ '__PLACEHOLDER:\s*(.+?)\s*' }}/);
+	return match ? match[1] : 'No description provided';
+}
+
+/**
+ * Creates a DynamicStructuredTool from an n8n node.
+ * @param {INodeType} node - The n8n node to convert.
+ * @param {IExecuteFunctions} ctx - The execution context.
+ * @param {INodeParameters} nodeParameters - The node parameters.
+ * @returns {DynamicStructuredTool} The created tool.
+ */
+export function createNodeAsTool(
+	node: INodeType,
+	ctx: IExecuteFunctions,
+	nodeParameters: INodeParameters,
+): DynamicStructuredTool {
+	// Find all placeholder values in the node parameters
+	const placeholderValues = traverseObject(nodeParameters);
+
+	// Generate Zod schema from placeholder values
+	const schemaObj: { [key: string]: z.ZodString } = {};
+	for (const [key, value] of placeholderValues.entries()) {
+		const description = extractPlaceholderDescription(value);
+		schemaObj[key] = z.string().describe(description);
+	}
+	const schema = z.object(schemaObj);
+
+	// Get the tool description from node parameters or use the default
+	const toolDescription = ctx.getNodeParameter(
+		'toolDescription',
+		0,
+		node.description.description,
+	) as string;
+	type GetNodeParameterMethod = IExecuteFunctions['getNodeParameter'];
+
+	const tool = new DynamicStructuredTool({
+		name: node.description.name,
+		description: toolDescription ? toolDescription : node.description.description,
```

Comment:
Is this redundant? `node.description.description` was passed as a fallback value in `ctx.getNodeParameter` call above
```suggestion
		description: toolDescription,
```

Reply:
I don't think this redundancy would be causing any issues, and I'd prefer to have this description in case user passes undefined or null via expression